lib/SPIRV/SPIRVReader.cpp

```diff
@@ -666,117 +666,106 @@ void SPIRVToLLVM::setLLVMLoopMetadata(const LoopInstType *LM,
-            for (auto &Op : PreviousIdxGroup->operands())
-              CurrentDepthOperands.push_back(Op);
-            if (CurrentDepthOperands.size() == 0)
-              CurrentDepthOperands.push_back(PreviousIdxGroup);
-            CurrentDepthOperands.push_back(CurrentDepthIdxGroup);
-            auto *JointIdxGroup =
-                llvm::MDNode::get(*Context, CurrentDepthOperands);
-            GEP->setMetadata(IdxGroupMDName, JointIdxGroup);
-          }
+    // Create index group metadata nodes - one per each array
+    // variables. Mark each GEP accessing a particular array variable
+    // into a corresponding index group
+    std::map<unsigned, std::vector<MDNode *>> SafelenIdxGroupMap;
+    for (auto &ArrayGEPIt : ArrayGEPMap) {
+      // Emit a distinct index group that will be referenced from
+      // llvm.loop.parallel_access_indices metadata
+      auto *CurrentDepthIdxGroup = llvm::MDNode::getDistinct(*Context, None);
+      unsigned Safelen = ArraySflnMap.find(ArrayGEPIt.first)->second;
+      SafelenIdxGroupMap[Safelen].push_back(CurrentDepthIdxGroup);
+
+      for (auto *GEP : ArrayGEPIt.second) {
+        StringRef IdxGroupMDName("llvm.index.group");
+        llvm::MDNode *PreviousIdxGroup = GEP->getMetadata(IdxGroupMDName);
+        if (!PreviousIdxGroup) {
+          GEP->setMetadata(IdxGroupMDName, CurrentDepthIdxGroup);
+          continue;
         }
 
-        for (auto &SflnIdxGroupIt : SafelenIdxGroupMap) {
-          auto *Name =
-              MDString::get(*Context, "llvm.loop.parallel_access_indices");
-          unsigned SflnValue = SflnIdxGroupIt.first;
-          llvm::Metadata *SafelenMDOp =
-              SflnValue ? ConstantAsMetadata::get(ConstantInt::get(
-                              Type::getInt32Ty(*Context), SflnValue))
-                        : nullptr;
-          std::vector<llvm::Metadata *> Parameters{Name};
-          for (auto *Node : SflnIdxGroupIt.second)
-            Parameters.push_back(Node);
-          if (SafelenMDOp)
-            Parameters.push_back(SafelenMDOp);
-          Metadata.push_back(llvm::MDNode::get(*Context, Parameters));
-        }
-        break;
-      }
-      default:
-        llvm_unreachable(
-            "Unexpected token in LoopControlExtendedConstrolsMask");
+        // If we're dealing with an embedded loop, it may be the case
+        // that GEP instructions for some of the arrays were already
+        // marked by the algorithm when it went over the outer level loops.
+        // In order to retain the IVDep information for each "loop
+        // dimension", we will mark such GEP's into a separate joined node
+        // that will refer to the previous levels' index groups AND to the
+        // index group specific to the current loop.
+        std::vector<llvm::Metadata *> CurrentDepthOperands;
+        for (auto &Op : PreviousIdxGroup->operands())
+          CurrentDepthOperands.push_back(Op);
+        if (CurrentDepthOperands.size() == 0)
```

Comment:
```suggestion
        if (CurrentDepthOperands.empty())
```

Reply:
Fixed this and other code-style leftovers with a separate commit